frontend/src/utils/account-send.ts

```diff
@@ -0,0 +1,144 @@
+import { isAddress, keccak256, toUtf8Bytes, BigNumber, hexZeroPad } from 'src/utils/ethers';
+import localforage from 'localforage';
+import { toAddress, lookupAddress } from 'src/utils/address';
+import { LOCALFORAGE_ACOUNT_SEND_KEY } from 'src/utils/constants';
+import { Web3Provider } from 'src/utils/ethers';
+
+type AccountData = {
+  address: string;
+  advancedMode: boolean;
+  checkbox: boolean;
+};
+
+type KeyData = {
+  count: number;
+  viewingKey: string;
+};
+
+type EncryptedData = {
+  encryptedAddress: string;
+};
+
+type AccountSendData = {
+  amount: string;
+  tokenAddress: string;
+  dateSent: Date;
+  hash: string;
+  userAddress: string;
+  recipientAddress: string;
+  checkbox?: boolean;
+  advancedMode?: boolean;
+};
+
+type DecryptedSendData = {
+  recipientId: string;
+};
+
+const accountData = ({ address, advancedMode, checkbox }: AccountData) => {
+  if (!isAddress(address)) {
+    throw new Error('Invalid address');
+  }
+  const advancedModeBit = advancedMode ? '1' : '0';
+  const checkboxBit = checkbox ? '1' : '0';
+  const addressHex = BigNumber.from(address).toHexString() + advancedModeBit + checkboxBit;
+
+  return BigNumber.from(hexZeroPad(addressHex, 32));
+};
+
+const encryptAccountData = ({ address, advancedMode, checkbox, count, viewingKey }: AccountData & KeyData) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${count}`));
+  const data = accountData({ address, advancedMode, checkbox });
+  const encryptedData = data.xor(key);
+  return hexZeroPad(encryptedData.toHexString(), 32);
+};
+
+export const decryptData = ({ viewingKey, count, encryptedAddress }: KeyData & EncryptedData) => {
+  const key = keccak256(toUtf8Bytes(`${viewingKey}${count}`));
+
+  const decryptedData = BigNumber.from(encryptedAddress).xor(key);
+  const hexData = decryptedData.toHexString();
+
+  const advancedMode = hexData.slice(-1);
+  const checkbox = hexData.slice(-2);
+  return {
+    advancedMode,
+    checkbox,
+    address: hexData.slice(0, -2),
+  };
+};
+
+export const storeSend = async ({
+  recipientAddress,
+  chainId,
+  advancedMode,
+  checkbox,
+  viewingKey,
+  amount,
+  tokenAddress,
+  hash,
+  userAddress,
+  provider,
+}: { chainId: number; viewingKey: string; provider: Web3Provider } & Omit<AccountSendData, 'dateSent'> &
```

Comment:
I feel like this destructuring of input args then manually combining types makes things harder to read. Is this just a personal preference, or is there another reason? Feels easier to just take one arg for each type you need in all these functions

Reply:
https://github.com/ScopeLift/umbra-protocol/pull/494/commits/67b1b5d4b80da11719360901bbfa04b1657517ca

Reply:
Should be remove here https://github.com/ScopeLift/umbra-protocol/pull/494/commits/dc610408e6e5907d67163b3b4e60bb3e59f84a8b

Reply:
Nice I like this a lot better, thanks! For `type StoreSendArgs`, should we use the existing types to reduce some code? e.g.

```typescript
type StoreSendArgs = {
  accountDataToEncrypt: AccountDataToEncrypt;
  unencryptedAccountSendData: UnencryptedAccountSendData;
  chainId: number;
  viewingKey: string;
  provider: Web3Provider;
};
```

I know you did it this way cause I mentioned the original concatenation of things into one giant type felt confusing, but this seems much more readable to me (maybe because this is what it'd look like in rust) (I see there's one param of `UnencryptedAccountSendData` we don't use but that seems ok)

Reply:
Yes, definitely can reuse those those and add the omit to the unecryptedAccountSendData. It would make it a little DRYer.

Reply:
Reorganized https://github.com/ScopeLift/umbra-protocol/pull/494/commits/52f182c6fa84bedf64f0f048a92c4a0c39ab0547

Reply:
love it, thank you 🫡